platform/analysis-api/src/com/intellij/openapi/roots/GeneratedSourcesFilter.java

```diff
@@ -16,16 +16,31 @@
 public abstract class GeneratedSourcesFilter {
   public static final ExtensionPointName<GeneratedSourcesFilter> EP_NAME = ExtensionPointName.create("com.intellij.generatedSourcesFilter");
 
+  /**
+   * Checks whether the given files is considered a generated source file in the context of the given project.
+   *
+   * This method can block, if a write action is currently pending.
+   *
+   * @return {@code true} if the file is considered a generated source, {@code false} otherwise.
+   */
   public static boolean isGeneratedSourceByAnyFilter(@NotNull VirtualFile file, @NotNull Project project) {
-    return ReadAction.compute(() -> {
-      if (project.isDisposed() || !file.isValid()) return false;
-      for (GeneratedSourcesFilter filter : EP_NAME.getExtensionList()) {
-        if (filter.isGeneratedSource(file, project)) {
-          return true;
-        }
-      }
-      return false;
-    });
+    return ReadAction.compute(() -> findFirstMatchingFilterNonBlocking(file, project) != null);
+  }
+
+  /**
+   * Returns the first source filter (in order of generated source filter extension traversal order) that considers this file
+   * to be generated. This method is non-blocking.
+   *
+   * @return the first {@code GeneratedSourcesFilter} that considered this file to be generated.
+   */
+  public static GeneratedSourcesFilter findFirstMatchingFilterNonBlocking(@NotNull VirtualFile file, @NotNull Project project) {
```

Comment:
Since it is not a private method now, it is better to either acquire read action or mark `@RequiresReadLock`

Reply:
Any opinion on which is better? I have originally explicitly made it non-blocking (hence the name), because I assumed that notification providers should do as little as possible. Would `@RequiresReadLock` be a better option in that case? Also, is there anything I would have to do to check this?

Reply:
> Also, is there anything I would have to do to check this?

Sorry, I'm not sure what you mean by that. `RequiresReadLock` throws an assertion if read access is not allowed

Well, I think it is better to mark it as `RequiresReadLock` and remove `NonBlocking` suffix

Reply:
Ah, I kind of assumed it's some static contract annotation and I would have to run something that checks this invariant. If it just inserts an assertions then yeah, I don't have to do anything else.

The method should now have the annotation.